chore(plugin-server): add ClickHouse inserter, remove KafkaTable

hazzadous opened the pull request:
NOTE: this PR was triggered by seeing additional KafkaTables introduced as a result of 
[live events](https://github.com/PostHog/posthog/pull/13019). It is slightly off piste but 
it's been bugging me for a while now and has a number of benefits.

Rather than using ClickHouse KafkaTable engine for inserting events into
ClickHouse, use the node ClickHouse client to do so. This avoids
coupling the database workload with data loading workloads. With this
setup we can control each independently of the other, i.e. if we want to
stop ingestion we do not have to play around with Database schema

NOTE: we create a new topic here, which we will consume from to insert into ClickHouse. The old KafkaTable will still consume from the existing topic, which should be exhausted, at which point either via async migration or otherwise we can clean up. TBC how this happens. 

TODO: 

 1. use Node streams API for inserts.
 4. send data directly to the appropriate replicas. At the moment we rely on the `writable_events` Distributed table to handle this, but we could reduce traffic if we do things a little smarter. But also YAGNI. We shall see, it'll probably be insignificant.
 5. At the moment I've limited this to just the app metrics and ingestion warning topics. Assuming these work out, we need to apply to all topics. I had to do some pretty annoying partition assignment stuff to work around that KafkaJS doesn't have an implementation of the range assignment strategy, and hence as 1. we still have some Kafka tables and 2. want to reuse consumer group `group1` such that the switchover is seamless (i.e. we use the correct offsets) we end up with a conflict in assignment strategy 😓 

For context on some issues we've had with KafkaTable, see:

1. https://github.com/PostHog/incidents-analysis/blob/master/2022-11-09-migration-stuck-us.md
2. https://github.com/PostHog/incidents-analysis/blob/master/2022-11-09-EU-MSK-outage.md

Aside from these, there have been some other times that there has been off behaviour with KafkaTables relating to but not limited deleting and recreating KafkaTables, tables seemingly stopping ingestion and needing to be recreated.

Aside from these issues, it also makes infrastructure and development more simple in that we no longer request to configure ClickHouse to contain the credentials and configuration for Kafka:

 1. we should no longer need to be concerned with the kafka bootstrap urls in the KafkaTable definition going out of sync with the Kafka instance you want to pull from (simply update env variables, as opposed to needing to perform a migration to remove and readd the KafkaTable with the new bootstrap urls, or doing so by hand)
 2. Changes to Kafka auth for instance no longer needs a reprovisioning of the ClickHouse configuration files.
 6. local development no longer needs the addition of a hard coded `/etc/hosts` entry for kafka 
 7. This change also makes it possible to use ClickHouse providers that do not support KafkaTable e.g. ClickHouse Cloud
 
## Problem

<!-- Who are we building for, what are their needs, why is this important? -->

## Changes

<!-- If there are frontend changes, please include screenshots. -->
<!-- If a reference design was involved, include a link to the relevant Figma frame! -->

👉 *Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

## How did you test this code?

<!-- Briefly describe the steps you took. -->
<!-- Include automated tests if possible, otherwise describe the manual testing routine. -->

Commit:
chore(plugin-server): add clickhouse inserter for events

Rather than using ClickHouse KafkaTable engine for inserting events into
ClickHouse, use the node ClickHouse client to do so. This avoids
coupling the database workload with data loading workloads. With this
setup we can control each independently of the other, i.e. if we want to
stop ingestion we do not have to play around with Database schemas.

This change currently doesn't:

 1. handle any bulk loading.
 2. handle tables other than the events table.
 3. attempt to optimize the insertion to particular nodes e.g. based on
    sharding key.

Commit:
Also remove the other KafkaTables

Commit:
fix typings

Commit:
fix typings

Commit:
fix typings

Commit:
update todo

Commit:
wip

Commit:
wip

Commit:
wip

Commit:
Remove test for KafkaTable, we no longer have them and instead insert via the plugin-server

Commit:
wip

Commit:
format

Commit:
chore(plugin-server): start ClickHouse consumer on jest test runs

Now that we don't use KafkaTable, and given that the tests assert things
about ClickHouse data, we need to ensure that the data reaches
ClickHouse.

Commit:
chore(plugins-server): remove uuid before inserting dlq event

The uuid does not exist in the dlq ClickHouse table and hence we recieve
an error on inserting.

Commit:
silence logging in jest tests

Commit:
limit to app metrics and ingestion warnings

Commit:
wip

Commit:
revert replicated sync test changes

Commit:
add back kafka/mv tables

hazzadous commented:
@yakkomajuri I added you as a reviewer here, but I'm updating to reduce the scope of this to just app metrics and ingestion warnings topics so have removed you for now.

Commit:
wip

Commit:
update custom assigner comment with TODO

xvello commented on plugin-server/src/main/ingestion-queues/batch-processing/each-batch-async-handlers.ts:
```diff
@@ -1,15 +1,27 @@
+            clickHouseEvent = JSON.parse(message.value.toString())
+            event = convertToIngestionEvent(clickHouseEvent)
+        } catch (error) {
+            return
```
Shall we send it to Sentry for debugging?

hazzadous replied:
I'll add as a TODO as this was a slightly tangential thing to get the tests working. I was thinking DLQ rather than Sentry.

hazzadous replied:
And ensure that things that would have previously been send to Sentry still go there 👍

hazzadous replied:
Ok, I'm sending to Sentry now

Commit:
chore: capture exceptions, send to Sentry

Commit:
chore: remove custom assigner, just use a separate consumer group

Previously I was trying to reuse the consumer group used in ClickHouse
KafkaTables. Instead I'm changing to simply use a separate consumer
group. Trying to reuse a group on only one table causes us to need to
align the consumer group partitioning strategy across group members.

Instead we will, in a followup commit, simply copy over the group
offsets to a new consumer group.

Commit:
chore: create new consumer group, copy offsets

Commit:
fix import

Commit:
fix tests

hazzadous commented on plugin-server/functional_tests/clickhouse-inserter.test.ts:
```diff
@@ -0,0 +1,123 @@
+
+const topics = [
+    // NOTE: for introducing the ClickHouse inserter consumer, we are initially
+    // only enabling for App Metrics and Ingestion Warnings. We will enable for
```
```suggestion
    // only enabling for App Metrics. We will enable for
```

Commit:
Update plugin-server/functional_tests/clickhouse-inserter.test.ts

Commit:
chore: add test for offset migration

Commit:
Merge branch 'feat/add-clickhouse-ingester' of https://github.com/PostHog/posthog into feat/add-clickhouse-ingester

Commit:
chore: remove unrelated changees

When I have the other topics using the inserter, we have some errors
inserting into ClickHouse due to unknown columns being inserted. As I've
now changed this to only use the inserter for the
`clickhouse_app_metrics` topic we don't need these changes.

Commit:
add back ingestion warnings create tables

Commit:
fix: add missing arg

Commit:
chore: refactor Kafka consumer tests

Have updated to not use the __iter__ consumer method and rather use the
poll method with a timeout.

Commit:
fix migration test

Commit:
fix consumer group name

Commit:
use correct topic for app metrics

Commit:
specify database qualified table names

Commit:
handle test case where topic empty

Commit:
chore: move offset migration to management command

Having this as a management command separate from ClickHouse migrations
feels a little less confusing.

Commit:
update comments

Commit:
move test

Commit:
wip

Commit:
wip

Commit:
wip

Commit:
wip

Commit:
wip

xvello commented on posthog/management/commands/migrate_clickhouse_consumer_offsets.py:
```diff
@@ -0,0 +1,109 @@
+    def add_arguments(self, parser):
+        parser.add_argument("--fake", action="store_true", help="Mark migrations as run without actually running them.")
+        parser.add_argument(
+            "--check", action="store_true", help="Mark migrations as run without actually running them."
```
Bad description?

hazzadous replied:
yes

xvello commented on posthog/management/commands/migrate_clickhouse_consumer_offsets.py:
```diff
@@ -0,0 +1,109 @@
+    def handle(self, *args, **options):
+        admin = KafkaAdminClient()
+
+        for topic, group in TOPICS_TO_MIGRATE:
```
nit: for a second wave of topic migrations, we won't want to re-migrate `clickhouse_app_metrics` again and rewind, so that should be a command param. But we can change that later on.

Commit:
chore: remove any idea of offset copying

Instead we will produce/consume to a new topic, leaving the existing
KafkaTable to consume from the existing topic.

TODO: can we put an async migration in this release to remove the topic
when fully consumed? Otherwise we do so for a subsequent release.

Commit:
update functional test

Commit:
wip

Commit:
wip

Commit:
Double write to Kafka/ClickHouse for app metrics

Commit:
fix sql

Commit:
fix shard issue

Commit:
fix distributed issue

xvello commented on plugin-server/src/main/ingestion-queues/batch-processing/each-batch-async-handlers.ts:
```diff
@@ -1,15 +1,35 @@
+            clickHouseEvent = JSON.parse(message.value.toString())
+            event = convertToIngestionEvent(clickHouseEvent)
+        } catch (error) {
+            Sentry.captureException(error)
```
nit: Let's capture the json if we managed to parse it:
```suggestion
            Sentry.captureException(error, { extra: { clickHouseEvent } })
```

xvello reviewed (commented):
Looks good, although I'm too much of a node n00b to approve for now.

Commit:
Update plugin-server/src/main/ingestion-queues/batch-processing/each-batch-async-handlers.ts

Co-authored-by: Xavier Vello <xavier@posthog.com>
